### PR TITLE
010 new view role detail

### PR DESCRIPTION
**Pull request by Joshua-Noble-Dev**

### CODE REVIEW

File: src/main/java/org/example/models/JobRole.java [hunk 1]
```diff
@@ -27,19 +30,95 @@ public class JobRole {
...
+        private int id;
+        private String roleName;
+        private String location;
+        private String capability;
+        private String band;
+        private Date closingDate;
+        private String status;
+        private String description;
+        private String responsibilities;
+        private String jobSpec;
```
Comment (joel-and) [inline src/main/java/org/example/models/JobRole.java:65]: Why remove the default constructor and add builder?

Comment (Joshua-Noble-Dev) [reply src/main/java/org/example/models/JobRole.java:65]: cause that was a work around to the 7+ parmeters linting check

**Review by joel-and — APPROVED** [review]

**Comment by joel-and** [comment]
Approved

File: src/main/java/org/example/controllers/JobRoleController.java [hunk 2]
```diff
@@ -32,4 +35,23 @@ public Response getAllJobRoles() {
...
+    @GET
+    @Path("/job-roles/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getJobRoleById(@PathParam("id") final int id) {
+        try {
+            return Response.ok().entity(
+                    jobRoleService.getJobRoleById(id)).build();
+        } catch (InvalidException e) {
+            return Response.status(
+                    Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
```
Comment (shaunganley) [inline src/main/java/org/example/controllers/JobRoleController.java]: shouldn't have a invalid exception if the user is not providing any data

File: src/main/java/org/example/daos/JobRoleDao.java [hunk 3]
```diff
@@ -20,22 +23,59 @@ public List<JobRole> getAllJobRoles(
...
+                        .build();
 
                 jobRolesList.add(role);
             }
 
         return jobRolesList;
     }
+    public JobRole getJobRoleById(final int id,
+                                  final Connection connection)
+        throws SQLException, DoesNotExistException, InvalidException {
```
Comment (shaunganley) [inline src/main/java/org/example/daos/JobRoleDao.java]: some of these exceptions not used

File: src/main/java/org/example/exceptions/Entity.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package org.example.exceptions;
+
+public enum Entity {
+    ROLEDETAIL("ID");
```
Comment (shaunganley) [inline src/main/java/org/example/exceptions/Entity.java]:
```suggestion
    ROLE("Role");
```

File: src/main/java/org/example/exceptions/InvalidException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package org.example.exceptions;
+
+public class InvalidException extends Throwable {
```
Comment (shaunganley) [inline src/main/java/org/example/exceptions/InvalidException.java]: not needed

File: src/main/java/org/example/models/JobRole.java [hunk 6]
```diff
@@ -6,6 +6,9 @@
 
 public class JobRole {
 
+    public JobRole() {
```
Comment (shaunganley) [inline src/main/java/org/example/models/JobRole.java:9]: needed?

File: src/main/java/org/example/models/JobRole.java [hunk 7]
```diff
@@ -27,19 +30,95 @@ public class JobRole {
...
+        this.band = builder.band;
+        this.closingDate = builder.closingDate;
+        this.status = builder.status;
+        this.description = builder.description;
+        this.responsibilities = builder.responsibilities;
+        this.jobSpec = builder.jobSpec;
+    }
+
+    // Static nested Builder class
+    public static class Builder {
```
Comment (shaunganley) [inline src/main/java/org/example/models/JobRole.java:55]: I like this, nice work

File: src/main/java/org/example/models/JobRole.java [hunk 8]
```diff
@@ -97,4 +176,32 @@ public String getStatus() {
...
+        return jobSpec;
+    }
+
+    public void setJobSpec(final String jobSpec) {
+        this.jobSpec = jobSpec;
+    }
+
+    public int compareTo(final JobRole o) {
+        return 0;
+    }
```
Comment (shaunganley) [inline src/main/java/org/example/models/JobRole.java]: doesn't work

File: src/main/java/org/example/services/JobRoleService.java [hunk 9]
```diff
@@ -19,7 +23,25 @@ public JobRoleService(final JobRoleDao roleDao,
...
+
+    public JobRole getJobRoleById(final int detailId)
+            throws SQLException, DoesNotExistException,
+            InvalidException {
+
+        JobRole jobRole = roleDao.getJobRoleById(detailId,
+                databaseConnector.getConnection());
+
+        if (detailId == 0) {
+            throw new InvalidException(Entity.ROLEDETAIL);
```
Comment (shaunganley) [inline src/main/java/org/example/services/JobRoleService.java]: is this a valid scenario? your dao doesn't return 0 from what I can see

File: src/test/java/integration/JobRoleIntegrationTest.java [hunk 10]
```diff
@@ -28,4 +29,17 @@ void getJobRoles_ShouldReturnListOfJobRoles() {
 
         Assertions.assertFalse(response.isEmpty());
     }
+
+    @Test
+    void getJobRoleById_shouldReturnJobRole() {
```
Comment (shaunganley) [inline src/test/java/integration/JobRoleIntegrationTest.java:34]: add another test with invalid id

**Review by shaunganley — CHANGES_REQUESTED** [review]
nice work, few small comments

File: src/main/java/org/example/models/JobRole.java (hunk 6 above)
Comment (Joshua-Noble-Dev) [reply src/main/java/org/example/models/JobRole.java:9]: yes very important but i wont say why :>

File: src/main/java/org/example/controllers/JobRoleController.java [hunk 11]
```diff
@@ -43,7 +43,7 @@ public Response getJobRoleById(@PathParam("id") final int id) {
         try {
             return Response.ok().entity(
                     jobRoleService.getJobRoleById(id)).build();
-        } catch (InvalidException e) {
+        } catch (IllegalArgumentException e) {
             return Response.status(
                     Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
```
Comment (shaunganley) [inline src/main/java/org/example/controllers/JobRoleController.java]: you don't need this. a GET shouldn't be able to return a bad request

File: src/main/java/org/example/exceptions/IllegalArgumentException.java [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package org.example.exceptions;
+
+public class IllegalArgumentException extends Throwable {
```
Comment (shaunganley) [inline src/main/java/org/example/exceptions/IllegalArgumentException.java]: not needed

File: src/main/java/org/example/services/JobRoleService.java [hunk 13]
```diff
@@ -31,15 +31,15 @@ public List<JobRoleResponse> getAllRoles()
...
-            InvalidException {
+            IllegalArgumentException {
 
         JobRole jobRole = roleDao.getJobRoleById(detailId,
                 databaseConnector.getConnection());
 
-        if (detailId == 0) {
-            throw new InvalidException(Entity.ROLEDETAIL);
+        if (detailId <= 0) {
+            throw new IllegalArgumentException(Entity.ROLE);
```
Comment (shaunganley) [inline src/main/java/org/example/services/JobRoleService.java]: not needed. if the provided id is invalid you can just return does not exist

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — APPROVED** [review]

**Pull request merged**